### ISSUE TITLE
Tree Shaking is not working

### ISSUE THREAD

**Issue by MananTank** [thirdweb-dev/js#459]
## The Problem

importing even the tiniest thing from SDK results in a TON of code being imported and results in an extremely huge bundle size

### Example

Let's say we add this line to our application: - 

```javascript
import { ChainId } from '@thirdweb-dev/sdk/evm';
```
( above-shown code is taken from thirdweb.com codebase)

<br />

`ChainId` is a tiny object and I would expect that only a little bit of JS will be imported from thirdweb-sdk in the final bundle - But that's not what happens!

This is what happens in Next.js & Vite

<br/>

# Next.js

Areas marked with an arrow are loaded because of the above-shown import

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/22043396/206845679-ca5df889-9ff0-4ab8-8630-fb7d67a05342.png">

<img width="500" alt="image" src="https://user-images.githubusercontent.com/22043396/206843431-4e6f3ffb-b966-47ee-a279-39183e0348c5.png">

<br/>
<br/>

# Vite ( vanilla JS )

JS chunks created after running the build command: 

<img width="1467" alt="image" src="https://user-images.githubusercontent.com/22043396/206841893-2adb83c1-3015-4b79-9796-f79327bb794a.png">

A LOT (200kB !!!) of unnecessary JS is bundled together and the below-shown chunks are loaded on the client

<img width="1117" alt="image" src="https://user-images.githubusercontent.com/22043396/206842133-72ff57d0-8894-44f4-adb3-4546e0c71798.png">

<br/>
<br/>

## What I Expect

Only the imported code - `ChainId` object should have been bundled - because It's a simple object that does not have any dependencies. 

If the tree shaking was working properly - only the below-shown object would be imported from `thirdweb-sdk` - nothing else 

<img width="200" alt="image" src="https://user-images.githubusercontent.com/22043396/206842274-a28da05d-84b3-4c91-b4bf-f9fa9315d0b3.png">

## What I Get 

A lot of unnecessary code (200kb+) gets bundled!

<br />

## Checkout these Repos to replicate the issue 

* Vite - https://github.com/MananTank/thirdweb-sdk-treeshaking-issue-vite
* Next - https://github.com/MananTank/thirdweb-sdk-next-issue

<br />

## What's happening here? 

I suspect there's something wrong with how the imports/exports are done or maybe how the package is built that is causing the tree shaking to not work properly. I highly recommend investigating this as soon as possible

<br />

## Impact 🔥 

Websites using thirdweb-sdk are loading a LOT more JS than required - including https://thirdweb.com/ 

<br />

**Comment by MananTank** [comment]
Looks like this issue is also in `@thirdweb-dev/react`

I did `import { ChainId } from '@thirdweb-dev/react';` and same thing happened as detailed above

**Comment by jnsdls** [comment]
Yep, this is a long standing issue that we've been tracking but that sadly isn't a simple fix.

The fundamental way we're exporting and importing files within the packages probably have something to do with it, we also have some circular dependencies inside the packages/sdk package that likely are to blame for a lot of it.

We'd love deeper research into what is going on and how we could potentially fix this.

With all that said this is also an industry wide problem, lots of external dependencies rely on poly filling huge native packages to work in the first place and so we end up inheriting a lot of that.

Further investigation and PRs definitely appreciated here!

**Comment by MananTank** [comment]
I would love to dig deep into this and figure out what's the issue here - I don't see why this can't be fixed.

Can you clarify what you mean by "polyfilling native packages" - Can you give some examples? - I can start the investigation from there

**Comment by jnsdls** [comment]
here are some things I would do to start investigating this (in no particular order):

1. the `packages/sdk` package has a _definite_ problem with dependency cycles, easy to find by turning the eslint rule for it back on here: https://github.com/thirdweb-dev/js/blob/4cdd0bd6348494a256d7c6a2bdf8f7b5c20f6877/packages/sdk/.eslintrc.cjs#L5
2. we use [preconstruct](https://preconstruct.tools/) as a build tool today, it utilizes rollup (2) under the hood, there may just be configurations that we are not taking advantage off - we could also potentially switch to using rollup (3?) directly if preconstruct is too limited for our usecase (other possible options: `tsup` (based on esbuild) / direct `esbuild`)
3. we currently do not have good "cleanliness" when it comes to which files contain which util functions etc, and our export paths are often similar to `export * from "./..."` in a bunch of `index.ts` files in folders etc, I have a hunch this interferes with proper tree shaking (also encourages problems like dependency cycles (see 1.))
4. we may just be bundling dependencies that we don't need to (can we define some things as externals?) - this won't help with tree shaking but overall bundle size potentially
5. lots of dependencies rely on native node modules like "stream" "crypto" "buffer" etc that then need to be polyfilled at *some* level since developers will want to utilize these libraries in web-frontends where these aren't available - I believe this polyfilling (often-times directly bundled into dependencies or done by us) likely leads to at least bundle bloat if not more tree shaking issues.

There might be more, in general I would try to experiment with `@thirdweb-dev/sdk` directly first - I would advise the use of `yarn push` at the root level of monorepo (relies on [yalc](https://github.com/wclr/yalc) under the hood) for quick iteration cycles with reproduction cases.

I expect this to be non-trivial work FWIW.

**Comment by franklovefrank** [comment]
@jnsdls I want to get familiar with thirdweb's codebase and this seems fun :) I will look into it & report my findings

**Comment by jnsdls** [comment]
this will be addressed in v4

**Comment by kienngo98** [comment]
Hi @MananTank, I want to give a hand on this topic, to reduce the bundle size of the packages. At the moment, importing just the `<ThirdwebProvider />` from the SDK will add ~700kb of code. Which places do you think we should be looking at?
![354471785_6352233681486209_6384358492889319716_n](https://github.com/thirdweb-dev/js/assets/26052673/ffa4efea-afcf-4793-81cf-8c7054ce66c2)

**Comment by kienngo98** [comment]
>  improves the bundle size by 160kB

Amazing!

**Comment by kien-ngo** [comment]
@MananTank Just an update for the `sdk`: I have managed to narrow down this culprit at: `packages\sdk\src\evm\core\sdk.ts`

It has ~46.4kB of unshakable code
